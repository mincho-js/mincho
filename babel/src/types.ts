import { NodePath, Node, types as t, PluginPass } from '@babel/core';
import { Scope } from '@babel/traverse';

export type PluginOptions = {
  result: [string, string];
  /**
   * @deprecated no longer used
   */
  path?: string;
};

export type PluginState = {
  dependentNodes: Set<{ loc: Node['loc']; node: Node }>;
  styledNodes: Array<{ isAlreadyExported: boolean; ident: string }>;
  // declarators: Array<t.VariableDeclarator>;
} & PluginPass & { opts: PluginOptions };

export type ProgramScope = Scope & {
  macaronData: {
    imports: Map<string, t.Identifier>;
    bindings: Array<NodePath<t.Node>>;
    nodes: Array<
      | {
          type: 'style';
          export: t.ExportNamedDeclaration;
          // node: t.CallExpression;
          name: string;
          shouldReExport: boolean;
        }
      | { type: 'binding'; node: t.Node }
    >;
    styles: Array<{
      shouldReExport: boolean;
      declaration: t.ExportNamedDeclaration;
      name: string;
    }>;
    cssFile: string;
  };
};
