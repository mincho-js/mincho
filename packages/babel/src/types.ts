import type { NodePath, PluginPass, types as t } from "@babel/core";
import type { Scope } from "@babel/traverse";

export type PluginOptions = {
  result: [string, string];
};

export type PluginState = PluginPass & {
  opts: PluginOptions;
};

export type ProgramScope = Scope & {
  minchoData: {
    imports: Map<string, t.Identifier>;
    bindings: Array<NodePath<t.Node>>;
    nodes: Array<t.Node>;
    cssFile: string;
  };
};

export type MinchoNode = ProgramScope["minchoData"]["nodes"][number];
