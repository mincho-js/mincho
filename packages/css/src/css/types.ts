import type { CSSRule } from "@mincho-js/transform-to-vanilla";
import type { Resolve } from "../types.js";

type IsOptional<T, K extends keyof T> =
  Record<string, never> extends Pick<T, K> ? true : false;
type IsRequired<T, K extends keyof T> =
  IsOptional<T, K> extends true ? false : true;

export type CSSRuleWith<T extends CSSRule> = {
  [K in keyof T as T[K] extends false ? never : K]: true extends T[K]
    ? K extends keyof CSSRule
      ? NonNullable<CSSRule[K]>
      : never
    : T[K];
} extends infer U
  ? Resolve<
      {
        [K in keyof U as IsRequired<U, K> extends true ? K : never]-?: U[K];
      } & {
        [K in keyof U as IsOptional<U, K> extends true ? K : never]?: U[K];
      }
    >
  : never;

// == Tests ====================================================================
// Ignore errors when compiling to CommonJS.
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node16', or 'nodenext'.
if (import.meta.vitest) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node16', or 'nodenext'.
  const { describe, it, assertType } = import.meta.vitest;

  describe.concurrent("CSSRuleWith Type Test", () => {
    function assertCssRuleWith<const T extends CSSRule>(
      variants: CSSRuleWith<T>
    ) {
      assertType<CSSRuleWith<T>>(variants);
      return variants;
    }

    it("Common CSSRuleWith Type", () => {
      assertCssRuleWith({
        color: "red",
        // @ts-expect-error: `size` is not css property.
        size: "small"
      });
    });

    it("Based on CSSRule Type", () => {
      type TestRule = {
        color: true;
        background?: true;
      };

      assertCssRuleWith<TestRule>({
        color: "red"
        // background is optional
      });
      assertCssRuleWith<TestRule>({
        color: "red",
        background: "blue"
      });

      // @ts-expect-error: `color` is required.
      assertCssRuleWith<TestRule>({
        background: "blue"
      });
      assertCssRuleWith<TestRule>({
        color: "red",
        // @ts-expect-error: backgroundColor is not included in the type.
        backgroundColor: "blue"
      });
    });

    it("Custom RuleWith Type", () => {
      type TestRule = {
        color: true;
        size: "medium" | "large";
        variant?: "outlined" | "filled";
      };

      assertCssRuleWith<TestRule>({
        color: "red",
        size: "medium"
      });
      assertCssRuleWith<TestRule>({
        color: "red",
        size: "medium",
        variant: "outlined"
      });

      // @ts-expect-error: `size` is required.
      assertCssRuleWith<TestRule>({
        color: "red"
      });
      assertCssRuleWith<TestRule>({
        color: "red",
        // @ts-expect-error: small is not included in `size` key.
        size: "small"
      });
      assertCssRuleWith<TestRule>({
        color: "red",
        size: "medium",
        // @ts-expect-error: `variant` only accepts "outlined" | "filled".
        variant: "ghost"
      });
    });
  });
}
