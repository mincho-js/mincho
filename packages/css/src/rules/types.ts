import type {
  ComplexCSSRule,
  CSSRule,
  ResolvedProperties,
  NonNullableString
} from "@mincho-js/transform-to-vanilla";

type Resolve<T> = {
  [Key in keyof T]: T[Key];
} & {};
export type ResolveComplex<T> =
  T extends Array<infer U> ? Array<Resolve<U>> : Resolve<T>;
type RemoveUndefined<T> = T extends undefined ? never : T;
export type RemoveUndefinedFromIntersection<T> = {
  [K in keyof T]: RemoveUndefined<T[K]>;
}[keyof T];

type UnionToIntersection<U> = (
  U extends unknown ? (k: U) => void : never
) extends (k: infer I) => void
  ? I
  : never;

type Primitive = string | number | boolean | null | undefined;
export type Serializable =
  | {
      [Key in string | number]: Primitive | Serializable;
    }
  | ReadonlyArray<Primitive | Serializable>;

type RecipeStyleRule = ComplexCSSRule | string;

export type VariantDefinitions = Record<string, RecipeStyleRule>;

type BooleanMap<T> = T extends "true" | "false" ? boolean : T;
export type ToggleVariantMap<ToggleVariants extends VariantDefinitions> = {
  [VariantGroup in keyof ToggleVariants]: {
    true: ToggleVariants[VariantGroup];
  };
};

export type VariantGroups = Record<string, VariantDefinitions>;
export type VariantObjectSelection<Variants extends VariantGroups> = {
  [VariantGroup in keyof Variants]?:
    | BooleanMap<keyof Variants[VariantGroup]>
    | undefined;
};
export type VariantToggleSelection<Variants extends VariantGroups> = {
  [VariantGroup in keyof Variants]: keyof Variants[VariantGroup] extends
    | "true"
    | "false"
    ? VariantGroup
    : never;
}[keyof Variants];
export type VariantSelection<Variants extends VariantGroups> =
  | VariantObjectSelection<Variants>
  | Array<VariantToggleSelection<Variants> | VariantObjectSelection<Variants>>;

export type VariantsClassNames<Variants extends VariantGroups> = {
  [P in keyof Variants]: {
    [PP in keyof Variants[P]]: string;
  };
};

export type PropTarget = keyof ResolvedProperties;

export type ComplexPropDefinition<PropKeys extends PropTarget | undefined> =
  | PropDefinition<PropKeys>
  | Array<PropKeys | PropDefinition<PropKeys>>;
export type PropDefinition<PropKeys extends PropTarget | undefined> =
  | PropDefinitionWithPropTarget<PropKeys>
  | PropDefinitionWithString<PropKeys>;

type PropDefinitionWithPropTarget<PropKeys extends PropTarget | undefined> = {
  [Key in Exclude<PropKeys, undefined>]?: {
    base: ResolvedProperties[Key];
  };
};
type PropDefinitionWithString<PropKeys extends PropTarget | undefined> = {
  [Key in NonNullableString]?:
    | {
        base?: ResolvedProperties[Exclude<PropKeys, undefined>];
        targets: PropKeys[];
      }
    | PropKeys[];
};

export type PropDefinitionOutput<
  T extends ComplexPropDefinition<PropTarget | undefined>
> = UnionToIntersection<
  T extends unknown[]
    ? PropDefinitionOutputElement<T[number]>
    : PropDefinitionOutputElement<T>
>;
type PropDefinitionOutputElement<DefinitionElement> =
  DefinitionElement extends string
    ? HandlePropTarget<DefinitionElement>
    : DefinitionElement extends { [key: string]: unknown }
      ? HandlePropDefinition<DefinitionElement>
      : never;

type HandlePropTarget<PropKeys extends string> = PropKeys extends PropTarget
  ? { [Key in PropKeys]: ResolvedProperties[Key] }
  : never;
type HandlePropDefinition<PropObject extends { [key: string]: unknown }> =
  UnionToIntersection<
    {
      [Key in keyof PropObject & string]: HandlePropDefinitionEntry<
        Key,
        PropObject[Key]
      >;
    }[keyof PropObject & string]
  >;
type HandlePropDefinitionEntry<
  Key extends string,
  PropValue
> = PropValue extends {
  targets: infer T;
}
  ? T extends unknown[]
    ? { [P in Key]: ResolvedProperties[Extract<T[number], PropTarget>] }
    : never
  : PropValue extends unknown[]
    ? { [P in Key]: ResolvedProperties[Extract<PropValue[number], PropTarget>] }
    : Key extends PropTarget
      ? { [P in Key]: ResolvedProperties[Key] }
      : never;

export type PatternResult<Variants extends VariantGroups> = {
  defaultClassName: string;
  variantClassNames: VariantsClassNames<Variants>;
  defaultVariants: VariantObjectSelection<Variants>;
  compoundVariants: Array<[VariantObjectSelection<Variants>, string]>;
};

export interface CompoundVariant<Variants extends VariantGroups> {
  variants: VariantSelection<Variants>;
  style: RecipeStyleRule;
}

export type ConditionalVariants<
  Variants extends VariantGroups | undefined,
  ToggleVariants extends VariantDefinitions | undefined
> = Variants extends undefined
  ? ToggleVariants extends undefined
    ? never
    : ToggleVariantMap<Exclude<ToggleVariants, undefined>>
  : ToggleVariants extends undefined
    ? Variants
    : Resolve<Variants & ToggleVariantMap<Exclude<ToggleVariants, undefined>>>;

export type PatternOptions<
  Variants extends VariantGroups | undefined,
  ToggleVariants extends VariantDefinitions | undefined,
  Props extends ComplexPropDefinition<PropTarget> | undefined
> = CSSRule & {
  base?: RecipeStyleRule;
  props?: Props;
  toggles?: ToggleVariants;
  variants?: Variants;
  defaultVariants?: VariantSelection<
    ConditionalVariants<Variants, ToggleVariants>
  >;
  compoundVariants?: Array<
    CompoundVariant<ConditionalVariants<Variants, ToggleVariants>>
  >;
};

export type RecipeClassNames<Variants extends VariantGroups> = {
  base: string;
  variants: VariantsClassNames<Variants>;
};

export type RuntimeFn<Variants extends VariantGroups> = ((
  options?: ResolveComplex<VariantSelection<Variants>>
) => string) & {
  variants: () => (keyof Variants)[];
  classNames: RecipeClassNames<Variants>;
};

export type RulesVariants<RuleFn extends RuntimeFn<VariantGroups>> =
  ResolveComplex<Parameters<RuleFn>[0]>;
export type RecipeVariants<RecipeFn extends RuntimeFn<VariantGroups>> =
  RulesVariants<RecipeFn>;

// == Tests ====================================================================
if (import.meta.vitest) {
  const { describe, it, assertType, expectTypeOf } = import.meta.vitest;

  describe.concurrent("ConditionVariants Type Test", () => {
    it("Conditional Type", () => {
      assertType<
        ConditionalVariants<{ color: { primary: string } }, undefined>
      >({ color: { primary: "className" } });

      assertType<ConditionalVariants<undefined, { disabled: string }>>({
        disabled: { true: "className" }
      });

      assertType<
        ConditionalVariants<
          { color: { primary: string } },
          { disabled: string }
        >
      >({
        color: { primary: "className1" },
        disabled: { true: "className" }
      });
    });
  });

  describe.concurrent("VariantSelection Type Test", () => {
    type ExampleVariants = {
      size: {
        small: string;
        large: string;
      };
      color: {
        red: string;
        blue: string;
      };
    };
    function assertSelectedVariant(
      variants: VariantSelection<ExampleVariants>
    ) {
      assertType<VariantSelection<ExampleVariants>>(variants);
      return variants;
    }

    it("Valid VariantSelection Type", () => {
      assertSelectedVariant({
        size: "small",
        color: "red"
      });
    });

    it("Invalid VariantSelection Type", () => {
      assertSelectedVariant({
        // @ts-expect-error: selected variant value is not included in `size` key.
        size: "medium", // error occurred here
        color: "red"
      });
    });
  });

  describe.concurrent("Compound Variants Type Test", () => {
    function assertCompoundVariants<Variants extends VariantGroups>(
      optionValue: CompoundVariant<Variants>
    ) {
      assertType(optionValue);
      return optionValue;
    }

    it("Valid CompoundVariant", () => {
      assertCompoundVariants({
        variants: {
          color: "brand",
          size: "small"
        },
        style: {
          fontSize: "16px"
        }
      });
    });

    it("Invalid CompoundVariant with default key", () => {
      assertCompoundVariants({
        // @ts-expect-error: selected variant key is not invalid
        variantss: {
          // ↑↑ error occurred here
          color: "brand",
          size: "small"
        },
        style: {
          fontSize: "16px"
        }
      });
    });
    it("Invalid CompoundVariant with style value", () => {
      assertCompoundVariants({
        variants: {
          color: "brand",
          size: "small"
        },
        style: {
          // @ts-expect-error: fonTsize does not exist in `ComplexCSSRule` Type
          fonTsize: "16px" // error occurred here
        }
      });
    });
  });

  describe.concurrent("RecipeClassNames Type Test", () => {
    function assertValidClassNames<Variants extends VariantGroups>(
      classNames: RecipeClassNames<Variants>
    ) {
      assertType(classNames);
      return classNames;
    }

    it("Valid ClassNames Type Check", () => {
      assertValidClassNames({
        base: "base-class",
        variants: {
          color: {
            brand: "color-brand-class",
            accent: "color-accent-class"
          },
          size: {
            small: "size-small-class",
            medium: "size-medium-class",
            large: "size-large-class"
          }
        }
      });
    });
  });

  describe.concurrent("RuntimeFn Type Test", () => {
    type TestVariantsType = {
      color: {
        brand: { color: string };
        accent: { color: string };
      };
      size: {
        small: { padding: number };
        medium: { padding: number };
        large: { padding: number };
      };
      outlined: {
        true: { border: string };
        false: { border: string };
      };
    };

    it("RuntimeFn Type Check", () => {
      type ExpectedResultType = RuntimeFn<TestVariantsType>;
      expectTypeOf<ExpectedResultType>().toBeFunction();
      expectTypeOf<ExpectedResultType["variants"]>().toBeFunction();
      expectTypeOf<ExpectedResultType["classNames"]>().toBeObject();

      const expectedResult = Object.assign(
        (_options?: ResolveComplex<VariantSelection<TestVariantsType>>) =>
          "my-class",
        {
          variants: () =>
            ["color", "size", "outlined"] satisfies (keyof TestVariantsType)[],
          classNames: {
            base: "basic-class",
            variants: {
              color: {
                brand: "color-brand",
                accent: "color-accent"
              },
              size: {
                small: "size-small",
                medium: "size-medium",
                large: "size-large"
              },
              outlined: {
                true: "outlined-true",
                false: "outlined-false"
              }
            }
          } satisfies RecipeClassNames<TestVariantsType>
        }
      );
      assertType<ExpectedResultType>(expectedResult);
    });
  });

  describe.concurrent("Types related to Rules", () => {
    function assertValidOptions<
      Variants extends VariantGroups | undefined = undefined,
      ToggleVariants extends VariantDefinitions | undefined = undefined,
      Props extends ComplexPropDefinition<PropTarget> | undefined = undefined
    >(options: PatternOptions<Variants, ToggleVariants, Props>) {
      assertType<PatternOptions<Variants, ToggleVariants, Props>>(options);
      return options;
    }

    it("Base Style PatternOptions", () => {
      // Flatten base style
      assertValidOptions({
        color: "red",
        backgroundColor: "gray",
        fontSize: 16
      });

      // Base style
      assertValidOptions({
        base: {
          color: "red",
          backgroundColor: "gray",
          fontSize: 16
        }
      });

      // Both
      assertValidOptions({
        color: "red",
        base: { backgroundColor: "gray", fontSize: 16 }
      });
    });

    it("Variant Style PatternOptions", () => {
      // Toggle style
      assertValidOptions({
        toggles: {
          disabled: { textDecoration: "line-through" },
          rounded: { borderRadius: 999 }
        }
      });

      // Variant style
      assertValidOptions({
        variants: {
          color: {
            brand: { color: "#FFFFA0" },
            accent: { color: "#FFE4B5" }
          },
          size: {
            small: { padding: 12 },
            medium: { padding: 16 },
            large: { padding: 24 }
          },
          outlined: {
            true: { border: "1px solid black" },
            false: { border: "1px solid transparent" }
          }
        }
      });

      // Both
      assertValidOptions({
        toggles: {
          disabled: { textDecoration: "line-through" },
          rounded: { borderRadius: 999 }
        },
        variants: {
          color: {
            brand: { color: "#FFFFA0" },
            accent: { color: "#FFE4B5" }
          },
          size: {
            small: { padding: 12 },
            medium: { padding: 16 },
            large: { padding: 24 }
          },
          outlined: {
            true: { border: "1px solid black" },
            false: { border: "1px solid transparent" }
          }
        }
      });
    });

    const toggleVariants = {
      disabled: { textDecoration: "line-through" },
      rounded: { borderRadius: 999 }
    } as const;
    const variants = {
      color: {
        brand: { color: "#FFFFA0" },
        accent: { color: "#FFE4B5" }
      },
      size: {
        small: { padding: 12 },
        medium: { padding: 16 },
        large: { padding: 24 }
      },
      outlined: {
        true: { border: "1px solid black" },
        false: { border: "1px solid transparent" }
      }
    } as const;

    it("Default Style PatternOptions", () => {
      // Toggle style
      assertValidOptions({
        defaultVariants: { disabled: true },
        toggles: toggleVariants
      });
      assertValidOptions({
        defaultVariants: ["disabled", { rounded: true }],
        toggles: toggleVariants
      });

      // Variant style
      assertValidOptions({
        defaultVariants: { color: "brand", outlined: true },
        variants
      });
      assertValidOptions({
        defaultVariants: ["outlined", { color: "brand" }],
        variants
      });

      // Both
      assertValidOptions({
        defaultVariants: {
          disabled: true,
          color: "brand",
          size: "medium",
          outlined: true
        },
        toggles: toggleVariants,
        variants
      });
      assertValidOptions({
        defaultVariants: [
          "disabled",
          "outlined",
          {
            color: "brand",
            size: "medium"
          }
        ],
        toggles: toggleVariants,
        variants
      });
    });

    it("Compound Style PatternOptions", () => {
      // Toggle style
      assertValidOptions({
        compoundVariants: [
          {
            variants: {
              disabled: true,
              rounded: true
            },
            style: {
              color: "red"
            }
          }
        ],
        toggles: toggleVariants
      });
      assertValidOptions({
        compoundVariants: [
          {
            variants: ["disabled", "rounded"],
            style: {
              color: "red"
            }
          }
        ],
        toggles: toggleVariants
      });

      // Variant style
      assertValidOptions({
        compoundVariants: [
          {
            variants: {
              color: "brand",
              outlined: true
            },
            style: {
              color: "red"
            }
          }
        ],
        variants
      });
      assertValidOptions({
        compoundVariants: [
          {
            variants: ["outlined", { color: "brand" }],
            style: {
              color: "red"
            }
          }
        ],
        variants
      });

      // Both
      assertValidOptions({
        compoundVariants: [
          {
            variants: {
              disabled: true,
              color: "brand",
              size: "medium",
              outlined: true
            },
            style: {
              color: "red"
            }
          }
        ],
        toggles: toggleVariants,
        variants
      });
      assertValidOptions({
        compoundVariants: [
          {
            variants: [
              "disabled",
              "outlined",
              {
                color: "brand",
                size: "medium"
              }
            ],
            style: {
              color: "red"
            }
          }
        ],
        toggles: toggleVariants,
        variants
      });
    });

    it("Props PatternOptions", () => {
      // Property Array
      assertValidOptions({
        props: ["color", "background"]
      });

      // Property with base value
      assertValidOptions({
        props: {
          background: { base: "red" }
        }
      });

      // Property with aliased
      assertValidOptions({
        props: {
          size: ["padding", "margin"]
        }
      });
      assertValidOptions({
        props: {
          size: { base: "3px", targets: ["padding", "margin"] }
        }
      });

      // Complex Props
      assertValidOptions({
        props: ["color", "background", { size: ["padding", "margin"] }]
      });
      assertValidOptions({
        props: [
          "color",
          {
            background: { base: "red" },
            contour: ["border", "outline"],
            size: { base: "3px", targets: ["padding", "margin"] }
          }
        ]
      });
    });

    it("Complex Style PatternOptions", () => {
      assertValidOptions({
        backgroundColor: "gray",
        base: { color: "red", fontSize: 16 },

        toggles: toggleVariants,
        variants,

        defaultVariants: {
          size: "small"
        },
        compoundVariants: [
          {
            variants: {
              color: "brand",
              size: "small"
            },
            style: {
              fontSize: "16px"
            }
          }
        ]
      });
    });
  });
}
