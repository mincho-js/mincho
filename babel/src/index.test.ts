import { transformSync } from '@babel/core';
import { macaronBabelPlugin } from './index';
import { PluginOptions } from './types';

function babelTransform(code: string) {
  const options: PluginOptions = { result: ['', ''] };
  const result = transformSync(code, {
    plugins: [[macaronBabelPlugin(), options]],
    presets: ['@babel/preset-typescript'],
    filename: 'test.tsx',
  });

  if (result === null || result.code === null)
    throw new Error(`Could not transform`);

  return { result: options.result, code: result.code };
}

test('export default style', () => {
  const { result, code } = babelTransform(`
    import { style } from '@macaron-css/core';
    
    export default style({
      color: 'red'
    })
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});

test('inside jsx expression', () => {
  const { result, code } = babelTransform(`
    import { style } from '@macaron-css/core';
    
    function App() {
      return <div class={style({
        color: 'red'
      })}>Hello</div>
    }

    console.log(red);
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});

test('hoists inline expression', () => {
  const { result, code } = babelTransform(`
    import { style } from '@macaron-css/core';
    const str = \`abc \${style({ color: 'red' })}\`
    console.log(str)
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});

test('hoists object property', () => {
  const { result, code } = babelTransform(`
    import { style } from '@macaron-css/core';
    const obj = {
      nested: {
        key: style({ color: 'red' })
      }
    }
    console.log(obj)
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});

test('hoists array member', () => {
  const { result, code } = babelTransform(`
    import { style } from '@macaron-css/core';
    const arr = [1, 2, style({ color: 'red' }), 4, style({ color: 'blue' })]
    console.log(arr)
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});

test.skip('extracts style function', () => {
  const { result, code } = babelTransform(`
    import { style } from '@macaron-css/core';
    
    const red = style({ color: 'red' });
    console.log(red);
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});

test('multiple variable declarators in one declaration', () => {
  const { result, code } = babelTransform(`
    import { style } from '@macaron-css/core';
    
    const red = style({ color: 'red' }),
      blue = style({ color: 'blue' }),
      green = style({ color: 'green' });

    console.log(red, blue, green);
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});

test('move bindings along with extracted style', () => {
  const { result, code } = babelTransform(`
    import { style } from '@macaron-css/core';

    const redColor = 'red';
    
    const red = style({ color: redColor });
    console.log(red);
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});

test('inside block scope', () => {
  const { result, code } = babelTransform(`
    import { style } from '@macaron-css/core';
    
    {
      const red = style({ color: 'red' });

      console.log(red);
    }
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});

test.skip('styled components', () => {
  const { result, code } = babelTransform(`
    import { styled } from '@macaron-css/solid';
    
    const Button = styled("button", {
      base: { color: 'red' }
    })

    console.log(Button)
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});

test('already exported', () => {
  const { result, code } = babelTransform(`
    import { style } from '@macaron-css/core';

    export const red = style({ color: 'red' });
    console.log(red);
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});

// doesn't work
test('hoisting same variable name in different scope', () => {
  const { result, code } = babelTransform(`
    import { style } from '@macaron-css/core';

    const red = style({ color: 'red' });
    console.log(red);

    function SomeComponent() {
      const red = style({ color: 'blue' });
      console.log(red)
    }
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});

test('array pattern hoisting', () => {
  const { result, code } = babelTransform(`
    import { createTheme } from '@macaron-css/core';

    const [themeClass, vars] = createTheme({
      colors: {
        brand: 'red'
      }
    });
    console.log(themeClass, vars);
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});

test.only('same binding in multiple declarations', () => {
  const { result, code } = babelTransform(`
    import { style } from '@macaron-css/core';

    const color = 'red';
    const foreground = style({ color });
    const background = style({ background: color });

    console.log(foreground, background)
  `);

  expect(result).toMatchSnapshot();
  expect(code).toMatchSnapshot();
});
