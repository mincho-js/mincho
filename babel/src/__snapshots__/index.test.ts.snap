// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`already exported 1`] = `
Array [
  "extracted_ndq7ki.css.ts",
  "import { style } from '@macaron-css/core';

const _red2 = style({
  color: 'red'
});

export var _red = style({
  color: 'red'
});",
]
`;

exports[`already exported 2`] = `
"import { _red as _red2 } from \\"extracted_ndq7ki.css.ts\\";
import { style } from '@macaron-css/core';
export { _red2 as red };
console.log(_red2);"
`;

exports[`array pattern hoisting 1`] = `
Array [
  "extracted_14e7vm3.css.ts",
  "import { createTheme } from '@macaron-css/core';
const [_themeClass2, _vars2] = createTheme({
  colors: {
    brand: 'red'
  }
});
export var [_themeClass2, _vars] = createTheme({
  colors: {
    brand: 'red'
  }
});",
]
`;

exports[`array pattern hoisting 2`] = `
"import { _vars as _vars2 } from \\"extracted_14e7vm3.css.ts\\";
import { _themeClass as _themeClass2 } from \\"extracted_14e7vm3.css.ts\\";
import { createTheme } from '@macaron-css/core';
console.log(_themeClass2, _vars2);"
`;

exports[`export default style 1`] = `
Array [
  "extracted_5qasdd.css.ts",
  "import { style } from '@macaron-css/core';
export var _$$unknown_macaron_identifier = style({
  color: 'red'
});",
]
`;

exports[`export default style 2`] = `
"import { _$$unknown_macaron_identifier as _$$unknown_macaron_identifier2 } from \\"extracted_5qasdd.css.ts\\";
import { style } from '@macaron-css/core';
export default _$$unknown_macaron_identifier2;"
`;

exports[`extracts style function 1`] = `
Array [
  "extracted_1sozav4.css.ts",
  "import { style } from '@macaron-css/core';

const _red2 = style({
  color: 'red'
});

export var _red = style({
  color: 'red'
});",
]
`;

exports[`extracts style function 2`] = `
"import { _red as _red2 } from \\"extracted_1sozav4.css.ts\\";
import { style } from '@macaron-css/core';
console.log(_red2);"
`;

exports[`hoisting same variable name in different scope 1`] = `
Array [
  "extracted_l7klua.css.ts",
  "import { style } from '@macaron-css/core';
export var _red = style({
  color: 'red'
});

const _red2 = style({
  color: 'red'
});

const _red4 = style({
  color: 'blue'
});

export var _red3 = style({
  color: 'blue'
});",
]
`;

exports[`hoisting same variable name in different scope 2`] = `
"import { _red3 as _red4 } from \\"extracted_l7klua.css.ts\\";
import { _red as _red2 } from \\"extracted_l7klua.css.ts\\";
import { style } from '@macaron-css/core';
console.log(_red2);

function SomeComponent() {
  console.log(_red4);
}"
`;

exports[`hoists array member 1`] = `
Array [
  "extracted_rkwtxb.css.ts",
  "import { style } from '@macaron-css/core';
const arr = [1, 2, style({
  color: 'red'
}), 4, style({
  color: 'blue'
})];
export var _$$macaron_arr = style({
  color: 'red'
});
export var _$$macaron_arr3 = style({
  color: 'blue'
});",
]
`;

exports[`hoists array member 2`] = `
"import { _$$macaron_arr3 as _$$macaron_arr4 } from \\"extracted_rkwtxb.css.ts\\";
import { _$$macaron_arr as _$$macaron_arr2 } from \\"extracted_rkwtxb.css.ts\\";
import { style } from '@macaron-css/core';
const arr = [1, 2, _$$macaron_arr2, 4, _$$macaron_arr4];
console.log(arr);"
`;

exports[`hoists inline expression 1`] = `
Array [
  "extracted_47l5xi.css.ts",
  "import { style } from '@macaron-css/core';
const str = \`abc \${style({
  color: 'red'
})}\`;
export var _$$macaron_str = style({
  color: 'red'
});",
]
`;

exports[`hoists inline expression 2`] = `
"import { _$$macaron_str as _$$macaron_str2 } from \\"extracted_47l5xi.css.ts\\";
import { style } from '@macaron-css/core';
const str = \`abc \${_$$macaron_str2}\`;
console.log(str);"
`;

exports[`hoists object property 1`] = `
Array [
  "extracted_2e1qtp.css.ts",
  "import { style } from '@macaron-css/core';
const obj = {
  nested: {
    key: style({
      color: 'red'
    })
  }
};
export var _$$macaron_key = style({
  color: 'red'
});",
]
`;

exports[`hoists object property 2`] = `
"import { _$$macaron_key as _$$macaron_key2 } from \\"extracted_2e1qtp.css.ts\\";
import { style } from '@macaron-css/core';
const obj = {
  nested: {
    key: _$$macaron_key2
  }
};
console.log(obj);"
`;

exports[`inside block scope 1`] = `
Array [
  "extracted_ugie7r.css.ts",
  "const _red2 = style({
  color: 'red'
});

import { style } from '@macaron-css/core';
export var _red = style({
  color: 'red'
});",
]
`;

exports[`inside block scope 2`] = `
"import { _red as _red2 } from \\"extracted_ugie7r.css.ts\\";
import { style } from '@macaron-css/core';
{
  console.log(_red2);
}"
`;

exports[`inside jsx expression 1`] = `
Array [
  "extracted_x1wc72.css.ts",
  "import { style } from '@macaron-css/core';

function App() {
  return <div class={style({
    color: 'red'
  })}>Hello</div>;
}

export var _$$macaron_App = style({
  color: 'red'
});",
]
`;

exports[`inside jsx expression 2`] = `
"import { _$$macaron_App as _$$macaron_App2 } from \\"extracted_x1wc72.css.ts\\";
import { style } from '@macaron-css/core';

function App() {
  return <div class={_$$macaron_App2}>Hello</div>;
}

console.log(red);"
`;

exports[`move bindings along with extracted style 1`] = `
Array [
  "extracted_7ip2jg.css.ts",
  "import { style } from '@macaron-css/core';
const redColor = 'red';

const _red2 = style({
  color: redColor
});

export var _red = style({
  color: redColor
});",
]
`;

exports[`move bindings along with extracted style 2`] = `
"import { _red as _red2 } from \\"extracted_7ip2jg.css.ts\\";
import { style } from '@macaron-css/core';
const redColor = 'red';
console.log(_red2);"
`;

exports[`multiple variable declarators in one declaration 1`] = `
Array [
  "extracted_93tetg.css.ts",
  "import { style } from '@macaron-css/core';

const _red2 = style({
  color: 'red'
}),
      blue = style({
  color: 'blue'
}),
      green = style({
  color: 'green'
});

const _red2 = style({
  color: 'red'
}),
      blue = style({
  color: 'blue'
}),
      green = style({
  color: 'green'
});

const _red2 = style({
  color: 'red'
}),
      blue = style({
  color: 'blue'
}),
      green = style({
  color: 'green'
});

export var _red = style({
  color: 'red'
});
export var _blue = style({
  color: 'blue'
});
export var _green = style({
  color: 'green'
});",
]
`;

exports[`multiple variable declarators in one declaration 2`] = `
"import { _green as _green2 } from \\"extracted_93tetg.css.ts\\";
import { _blue as _blue2 } from \\"extracted_93tetg.css.ts\\";
import { _red as _red2 } from \\"extracted_93tetg.css.ts\\";
import { style } from '@macaron-css/core';
console.log(_red2, _blue2, _green2);"
`;

exports[`same binding in multiple declarations 1`] = `
Array [
  "extracted_s8bnpw.css.ts",
  "import { style } from '@macaron-css/core';
const color = 'red';
const background = style({
  background: color
});
export var _foreground = style({
  color
});

const _foreground2 = style({
  color
});

export var _background = style({
  background: color
});",
]
`;

exports[`same binding in multiple declarations 2`] = `
"import { _background as _background2 } from \\"extracted_s8bnpw.css.ts\\";
import { _foreground as _foreground2 } from \\"extracted_s8bnpw.css.ts\\";
import { style } from '@macaron-css/core';
const color = 'red';
console.log(_foreground2, _background2);"
`;

exports[`styled components 1`] = `
Array [
  "extracted_304vki.css.ts",
  "",
]
`;

exports[`styled components 2`] = `
"import { styled } from '@macaron-css/solid';
const Button = styled(\\"button\\", {
  base: {
    color: 'red'
  }
});
console.log(Button);"
`;
